prog/postgres/postgres_server_nexus.rb

```diff
@@ -0,0 +1,176 @@
+
+      vm.sshable.cmd("sudo common/bin/add_to_fstab #{device_path} /dat ext4 defaults 0 0")
+      vm.sshable.cmd("sudo mount #{device_path} /dat")
+
+      hop_install_postgres
+    when "Failed", "NotStarted"
+      device_path = vm.vm_storage_volumes.find { _1.boot == false }.device_path.shellescape
+      vm.sshable.cmd("common/bin/daemonizer 'sudo mkfs --type ext4 #{device_path}' format_disk")
+    end
+
+    nap 5
+  end
+
+  label def install_postgres
+    case vm.sshable.cmd("common/bin/daemonizer --check install_postgres")
+    when "Succeeded"
+      hop_refresh_certificates
+    when "Failed", "NotStarted"
+      vm.sshable.cmd("common/bin/daemonizer 'sudo postgres/bin/install_postgres' install_postgres")
+    end
+
+    nap 5
+  end
+
+  label def refresh_certificates
+    vm.sshable.cmd("sudo -u postgres tee /var/lib/postgresql/16/main/server.crt > /dev/null", stdin: postgres_server.resource.server_cert)
+    vm.sshable.cmd("sudo -u postgres tee /var/lib/postgresql/16/main/server.key > /dev/null", stdin: postgres_server.resource.server_cert_key)
+    vm.sshable.cmd("sudo -u postgres chmod 600 /var/lib/postgresql/16/main/server.key")
+    vm.sshable.cmd("sudo -u postgres pg_ctlcluster 16 main reload")
+
+    when_initial_provisioning_set? do
+      hop_configure
+    end
+
+    hop_wait
+  end
+
+  label def configure
+    case vm.sshable.cmd("common/bin/daemonizer --check configure")
+    when "Succeeded"
+      when_initial_provisioning_set? do
+        hop_update_superuser_password
+      end
+
+      hop_wait
+    when "Failed", "NotStarted"
+      configure_hash = postgres_server.configure_hash
+      vm.sshable.cmd("common/bin/daemonizer 'sudo postgres/bin/configure' configure", stdin: JSON.generate(configure_hash))
+    end
+
+    nap 5
+  end
+
+  label def update_superuser_password
+    encrypted_password = DB.synchronize do |conn|
+      # This uses PostgreSQL's PQencryptPasswordConn function, but it needs a connection, because
+      # the encryption is made by PostgreSQL, not by control plane. We use our own control plane
+      # database to do the encryption.
+      conn.encrypt_password(postgres_server.resource.superuser_password, "postgres", "scram-sha-256")
+    end
+    commands = <<SQL
+BEGIN;
+SET LOCAL log_statement = 'none';
+ALTER ROLE postgres WITH PASSWORD #{DB.literal(encrypted_password)};
+COMMIT;
+SQL
+    vm.sshable.cmd("sudo -u postgres psql", stdin: commands)
+
+    when_initial_provisioning_set? do
+      hop_restart
+    end
+    hop_wait
+  end
+
+  label def restart
+    vm.sshable.cmd("sudo postgres/bin/restart")
+
+    when_initial_provisioning_set? do
+      decr_initial_provisioning
+    end
+    hop_wait
+  end
+
+  label def wait
+    when_refresh_certificates_set? do
+      hop_refresh_certificates
+    end
+
+    nap 30
+  end
+
+  label def destroy
+    decr_destroy
+
+    vm.private_subnets.each { _1.incr_destroy }
```

Comment:
Should we wait the vm destroy?

https://github.com/ubicloud/ubicloud/pull/555#discussion_r1319776075